File: dashboard/src/lib/board-schema.ts
// SOURCE OF TRUTH: crates/agentum-core/src/board_schema.rs::required_fields_for.
// Keep in sync. The server is authoritative — a drift here degrades UX (the
// form lets through a payload the server then rejects) but never corrupts
// data: the dashboard reacts to the 400's {missing, status} shape.

/// Field keys the server uses in its `missing[]` payload. The
/// `session_id_or_comment` entry is synthetic — it represents the `done`
/// transition's OR-clause (either `session_id IS NOT NULL` OR `>=1 row in
/// board_comments`). The client can't know about existing comments
/// authoritatively (no per-card "did anyone comment" probe), so the
/// dialog only checks `session_id` for the `done` gate and falls back
/// to the server's 400 + missing[] hint if the comment fallback would
/// have rescued the row.
export type RequiredField =
  | 'title'
  | 'lbl'
  | 'workdir'
  | 'tool'
  | 'claimed_by'
  | 'session_id_or_comment';

/// Empty array => no gate (custom columns passthrough).
export function requiredFieldsFor(status: string): RequiredField[] {
  switch (status) {
    case 'todo':
      return ['title', 'lbl'];
    case 'doing':
      return ['title', 'lbl', 'workdir', 'tool', 'claimed_by'];
    case 'done':
      return ['title', 'lbl', 'session_id_or_comment'];
    default:
      return [];
  }
}

/// Snapshot the validator needs. Mirrors the Rust `TransitionCtx`.
export interface TransitionCtx {
  title?: string | null;
  lbl?: string | null;
  workdir?: string | null;
  tool?: string | null;
  claimed_by?: string | null;
  session_id?: string | null;
  /** Only available authoritatively on the server. Client-side this stays
   *  false; the server-side 400 carries the truth when the dialog needs
   *  to react. */
  has_comment?: boolean;
}

function isSet(v: string | null | undefined): boolean {
  return typeof v === 'string' && v.trim().length > 0;
}

/// Mirrors `agentum_core::validate_transition`. Returns the list of
/// missing keys for the target column, or an empty array on pass.
/// Custom columns (anything outside the matrix) always pass.
export function validateTransition(
  targetStatus: string,
  ctx: TransitionCtx
): RequiredField[] {
  const required = requiredFieldsFor(targetStatus);
  if (required.length === 0) return [];

  const missing: RequiredField[] = [];
  for (const field of required) {
    let present = false;
    switch (field) {
      case 'title':
        present = isSet(ctx.title);
        break;
      case 'lbl':
        present = isSet(ctx.lbl);
        break;
      case 'workdir':
        present = isSet(ctx.workdir);
        break;
      case 'tool':
        present = isSet(ctx.tool);
        break;
      case 'claimed_by':
        present = isSet(ctx.claimed_by);
        break;
      case 'session_id_or_comment':
        // Comment-existence isn't reachable from the client. Server
        // remains authoritative for the OR-clause's right side.
        present = isSet(ctx.session_id) || ctx.has_comment === true;
        break;
    }
    if (!present) missing.push(field);
  }
  return missing;
}

/// Display-friendly label for a `RequiredField`. Used by the dialog's
/// "Required for *<status>*" hint and the snap-back toast.
export function requiredFieldLabel(field: RequiredField): string {
  switch (field) {
    case 'title':
      return 'title';
    case 'lbl':
      return 'label';
    case 'workdir':
      return 'workdir';
    case 'tool':
      return 'agent';
    case 'claimed_by':
      return 'claim';
    case 'session_id_or_comment':
      return 'session or a comment';
  }
}

/// Parse the server's `400 { missing, status }` body. Returns `null`
/// when the body isn't the expected shape — caller should fall back to
/// the generic error path.
export interface GateRejection {
  missing: RequiredField[];
  status: string;
}

export function parseGateRejection(body: unknown): GateRejection | null {
  if (typeof body !== 'object' || body === null) return null;
  const b = body as Record<string, unknown>;
  const missing = b.missing;
  const status = b.status;
  if (!Array.isArray(missing) || typeof status !== 'string') return null;
  // Only adopt known keys — defensive against the server adding a new
  // synthetic field before the dashboard catches up.
  const filtered = missing.filter(
    (k): k is RequiredField =>
      k === 'title' ||
      k === 'lbl' ||
      k === 'workdir' ||
      k === 'tool' ||
      k === 'claimed_by' ||
      k === 'session_id_or_comment'
  );
  return { missing: filtered, status };
}

File: crates/agentum-core/src/board_schema.rs
//! Per-status required-field matrix for board cards.
//!
//! Slice 1: matrix is a compile-time const. Slice 2 adds per-server
//! overrides on top — see `agentum-server::rules` for the lookup glue.
//! Custom columns (anything not in the match arms below) bypass the
//! const gate but can be opted in via overrides.

use serde::{Deserialize, Serialize};

/// Fields a card must satisfy to *enter* the column. The `done` OR-clause
/// (`session_id` OR `>=1 comment`) is encoded as a single synthetic field
/// `SessionOrComment`; the validator resolves the disjunction.
///
/// Serde renames are explicit per variant so the wire vocabulary matches
/// `as_missing_key()` exactly — `rename_all = "snake_case"` would not
/// produce `session_id_or_comment` from `SessionOrComment`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RequiredField {
    #[serde(rename = "title")]
    Title,
    #[serde(rename = "lbl")]
    Lbl,
    #[serde(rename = "workdir")]
    Workdir,
    #[serde(rename = "tool")]
    Tool,
    #[serde(rename = "claimed_by")]
    ClaimedBy,
    #[serde(rename = "session_id_or_comment")]
    SessionOrComment,
}

impl RequiredField {
    /// JSON-array key for the `missing[]` payload returned on a 400.
    /// `&'static str` so callers don't allocate.
    pub fn as_missing_key(self) -> &'static str {
        match self {
            Self::Title => "title",
            Self::Lbl => "lbl",
            Self::Workdir => "workdir",
            Self::Tool => "tool",
            Self::ClaimedBy => "claimed_by",
            Self::SessionOrComment => "session_id_or_comment",
        }
    }

    /// Inverse of [`as_missing_key`]. Returns `None` for unknown strings so
    /// the store can skip-and-warn on rows that pin a removed variant
    /// (forward-compat policy — see the architecture file's risk #3).
    pub fn from_missing_key(s: &str) -> Option<Self> {
        match s {
            "title" => Some(Self::Title),
            "lbl" => Some(Self::Lbl),
            "workdir" => Some(Self::Workdir),
            "tool" => Some(Self::Tool),
            "claimed_by" => Some(Self::ClaimedBy),
            "session_id_or_comment" => Some(Self::SessionOrComment),
            _ => None,
        }
    }
}

/// Snapshot the validator needs. Built by the handler from the existing
/// row (PATCH) or zeroed (POST), then merged with the incoming patch.
#[derive(Debug, Clone, Default)]
pub struct TransitionCtx<'a> {
    pub title: Option<&'a str>,
    pub lbl: Option<&'a str>,
    pub workdir: Option<&'a str>,
    pub tool: Option<&'a str>,
    pub claimed_by: Option<&'a str>,
    pub session_id: Option<&'a str>,
    pub has_comment: bool,
}

/// Empty slice => no gate (custom columns passthrough).
pub fn required_fields_for(status: &str) -> &'static [RequiredField] {
    use RequiredField::*;
    match status {
        "todo" => &[Title, Lbl],
        "doing" => &[Title, Lbl, Workdir, Tool, ClaimedBy],
        "done" => &[Title, Lbl, SessionOrComment],
        _ => &[],
    }
}

/// Validate `ctx` against an arbitrary required-field slice. The caller
/// chooses the source (const matrix or DB override); this function knows
/// nothing about column names. Returns `Err(missing_keys)` on gate failure.
pub fn validate_against(
    required: &[RequiredField],
    ctx: &TransitionCtx<'_>,
) -> Result<(), Vec<&'static str>> {
    if required.is_empty() {
        return Ok(());
    }

    let mut missing: Vec<&'static str> = Vec::new();
    for field in required {
        let present = match field {
            RequiredField::Title => is_set(ctx.title),
            RequiredField::Lbl => is_set(ctx.lbl),
            RequiredField::Workdir => is_set(ctx.workdir),
            RequiredField::Tool => is_set(ctx.tool),
            RequiredField::ClaimedBy => is_set(ctx.claimed_by),
            // OR-clause: either an attached session OR at least one
            // comment on the parent row. Preserves the manual-close path
            // ("won't fix", "dup of AG-12") for cards that never ran an
            // agent session.
            RequiredField::SessionOrComment => is_set(ctx.session_id) || ctx.has_comment,
        };
        if !present {
            missing.push(field.as_missing_key());
        }
    }
    if missing.is_empty() {
        Ok(())
    } else {
        Err(missing)
    }
}

/// Returns `Err(missing_keys)` on gate failure, `Ok(())` on pass or
/// custom-column passthrough. Thin shim over [`validate_against`] that
/// pins the required slice to the slice-1 const matrix — kept so slice-1
/// callsites and tests don't change.
pub fn validate_transition(
    target_status: &str,
    ctx: &TransitionCtx<'_>,
) -> Result<(), Vec<&'static str>> {
    validate_against(required_fields_for(target_status), ctx)
}

/// Treat `None` and whitespace-only strings as absent. Trim is the
/// minimum sanity check — the rest is up to the wire layer.
fn is_set(v: Option<&str>) -> bool {
    matches!(v, Some(s) if !s.trim().is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ctx_todo_pass<'a>() -> TransitionCtx<'a> {
        TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            ..Default::default()
        }
    }

    #[test]
    fn todo_pass() {
        assert!(validate_transition("todo", &ctx_todo_pass()).is_ok());
    }

    #[test]
    fn todo_fail_missing_lbl() {
        let ctx = TransitionCtx {
            title: Some("t"),
            ..Default::default()
        };
        let err = validate_transition("todo", &ctx).unwrap_err();
        assert_eq!(err, vec!["lbl"]);
    }

    #[test]
    fn doing_pass() {
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            workdir: Some("/tmp"),
            tool: Some("claude"),
            claimed_by: Some("alice"),
            ..Default::default()
        };
        assert!(validate_transition("doing", &ctx).is_ok());
    }

    #[test]
    fn doing_fail_missing_three_fields_in_order() {
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            ..Default::default()
        };
        let err = validate_transition("doing", &ctx).unwrap_err();
        // Order matters: the validator walks `required_fields_for` left
        // to right, so the missing[] array always has a deterministic
        // shape the dashboard can rely on.
        assert_eq!(err, vec!["workdir", "tool", "claimed_by"]);
    }

    #[test]
    fn done_pass_via_session() {
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            session_id: Some("abc"),
            ..Default::default()
        };
        assert!(validate_transition("done", &ctx).is_ok());
    }

    #[test]
    fn done_pass_via_comment() {
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            has_comment: true,
            ..Default::default()
        };
        assert!(validate_transition("done", &ctx).is_ok());
    }

    #[test]
    fn done_fail_neither_session_nor_comment() {
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            ..Default::default()
        };
        let err = validate_transition("done", &ctx).unwrap_err();
        assert_eq!(err, vec!["session_id_or_comment"]);
    }

    #[test]
    fn custom_column_passthrough() {
        // Any status not in the matrix returns Ok regardless of ctx —
        // user-introduced columns like `blocked` or `review` keep working
        // without a code change.
        let ctx = TransitionCtx::default();
        assert!(validate_transition("blocked", &ctx).is_ok());
        assert!(validate_transition("review", &ctx).is_ok());
        assert!(validate_transition("anything-goes", &ctx).is_ok());
    }

    #[test]
    fn whitespace_only_is_absent() {
        // Defensive: `Some("   ")` is treated as missing so a client
        // can't sneak past the gate with an empty payload.
        let ctx = TransitionCtx {
            title: Some("   "),
            lbl: Some("feat"),
            ..Default::default()
        };
        let err = validate_transition("todo", &ctx).unwrap_err();
        assert_eq!(err, vec!["title"]);
    }

    #[test]
    fn from_missing_key_roundtrip_all_variants() {
        // Every known variant must round-trip through the
        // as_missing_key / from_missing_key pair — the store relies on
        // this symmetry to parse JSON rows back into typed enums.
        for f in [
            RequiredField::Title,
            RequiredField::Lbl,
            RequiredField::Workdir,
            RequiredField::Tool,
            RequiredField::ClaimedBy,
            RequiredField::SessionOrComment,
        ] {
            assert_eq!(RequiredField::from_missing_key(f.as_missing_key()), Some(f));
        }
    }

    #[test]
    fn from_missing_key_unknown_is_none() {
        // Unknown strings return None so the store can skip-and-warn
        // instead of failing the whole row.
        assert!(RequiredField::from_missing_key("wat").is_none());
        assert!(RequiredField::from_missing_key("").is_none());
        assert!(RequiredField::from_missing_key("TITLE").is_none()); // case-sensitive
    }

    #[test]
    fn serde_roundtrip_uses_wire_strings() {
        // The wire format must produce `"session_id_or_comment"`, not
        // `"SessionOrComment"` (the variant ident) or anything else.
        let json = serde_json::to_string(&RequiredField::SessionOrComment).unwrap();
        assert_eq!(json, "\"session_id_or_comment\"");
        let back: RequiredField = serde_json::from_str("\"title\"").unwrap();
        assert_eq!(back, RequiredField::Title);
    }

    #[test]
    fn validate_against_with_explicit_slice() {
        // Smoke: pass an explicit required slice that diverges from the
        // const matrix and confirm the validator honors it. Mirrors how
        // the override path in slice 2 calls this.
        let ctx = TransitionCtx {
            title: Some("t"),
            lbl: Some("feat"),
            ..Default::default()
        };
        // Empty slice => unconditional pass.
        assert!(validate_against(&[], &ctx).is_ok());
        // Title-only requirement => pass.
        assert!(validate_against(&[RequiredField::Title], &ctx).is_ok());
        // Demand workdir on a ctx that has none => fail.
        let err = validate_against(&[RequiredField::Workdir], &ctx).unwrap_err();
        assert_eq!(err, vec!["workdir"]);
    }
}
